### PR TITLE
Ability to enable/disable attachment upload

### PR DESCRIPTION
**Pull request by ce-simsoft**
Improves https://github.com/NCI-Agency/anet/pull/4131

Attachment feature can be enabled/disabled.

Closes [AB#894](https://dev.azure.com/ncia-anet/2aa083a5-af3d-44e1-8c7b-6e9e6b124d91/_workitems/edit/894)

#### User changes
- none

#### Superuser changes
- none

#### Admin changes
- Attachment feature can be enabled/disabled via dictionary

#### System admin changes
- [x] anet.yml or anet-dictionary.yml needs change
- [ ] db needs migration
- [ ] documentation has changed
- [ ] graphql schema has changed

### Checklist
  - [x] Described the user behavior in PR body
  - [x] Referenced/updated all related issues
  - [x] commits follow a `repo#issue: Title` title format and [these 7 rules](https://chris.beams.io/posts/git-commit/)
  - [x] commits have a [clean history](https://epage.github.io/dev/commits/), otherwise PR may be squash-merged
  - [x] Added and/or updated unit tests
  - [x] Added and/or updated e2e tests
  - [ ] Added and/or updated data migrations
  - [ ] Updated documentation
  - [x] Resolved all build errors and warnings
  - [ ] Opened debt issues for anything not resolved here

### CODE REVIEW

File: src/main/java/mil/dds/anet/resources/AttachmentResource.java [hunk 1]
```diff
@@ -70,6 +70,13 @@ public List<Attachment> getRelatedObjectAttachments(
   @GraphQLMutation(name = "createAttachment")
   public String createAttachment(@GraphQLRootContext Map<String, Object> context,
       @GraphQLArgument(name = "attachment") Attachment attachment) {
+    final Map<String, Object> attachmentSettings = getAttachmentSettings();
+    final Boolean attachmentDisabled = (Boolean) attachmentSettings.get("featureDisabled");
+
+    if (attachmentDisabled) {
+      throw new WebApplicationException("Attachment feature is disabled", Status.FORBIDDEN);
+    }
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/resources/AttachmentResource.java:78]: You could extract this into a function and call that in both places.

File: client/src/pages/reports/Show.js [hunk 2]
```diff
@@ -673,29 +673,31 @@ const ReportShow = ({ setSearchQuery, pageDispatchers }) => {
...
-                          key={index}
-                          attachment={attachment}
-                          index={index}
-                          edit={false}
-                        />
-                      ))}
-                    </div>
-                  }
-                />
+                {Settings.fields.attachment.featureDisabled && (
```
Comment (gjvoosten) [inline client/src/pages/reports/Show.js:676]: Shouldn't this be `!Settings.fields.attachment.featureDisabled && (…` i.e. the inverse condition?

File: client/tests/webdriver/baseSpecs/showReport.spec.js [hunk 3]
```diff
@@ -34,16 +34,20 @@ describe("Show report page", () => {
   describe("WHen on the show page of a report with attachment(s)", () => {
     it("We should see a container for Attachment List", async() => {
       // Attachment container
-      await (await ShowReport.getAttachments()).waitForExist()
-      await (await ShowReport.getAttachments()).waitForDisplayed()
+      if (await (await ShowReport.getAttachments()).isExisting()) {
+        await (await ShowReport.getAttachments()).waitForExist()
+        await (await ShowReport.getAttachments()).waitForDisplayed()
+      }
```
Comment (gjvoosten) [inline client/tests/webdriver/baseSpecs/showReport.spec.js:40]: Why is this now a conditional check?

File: client/tests/webdriver/baseSpecs/showReport.spec.js [hunk 4]
```diff
@@ -34,16 +34,20 @@ describe("Show report page", () => {
...
-      expect(await await ShowReport.getFileData()).to.be.equal(
-        "test_at…\n12.0 KB"
-      )
+      if (await (await ShowReport.getCard()).isExisting()) {
+        await (await ShowReport.getCard()).waitForExist()
+        await (await ShowReport.getCard()).waitForDisplayed()
+        expect(await await ShowReport.getFileData()).to.be.equal(
+          "test_at…\n12.0 KB"
+        )
+      }
```
Comment (gjvoosten) [inline client/tests/webdriver/baseSpecs/showReport.spec.js:50]: Why is this now a conditional check?

File: src/test/java/mil/dds/anet/test/resources/AttachmentResourceTest.java [hunk 5]
```diff
@@ -34,181 +33,193 @@ public class AttachmentResourceTest extends AbstractResourceTest {
...
-    failAttachmentCreate(adminMutationExecutor, failedAttachmentInput);
-
-    if (userUploadDisabled) {
-      // Fail attachment create with any user other than admin
-      final MutationExecutor erinMutationExecutor =
-          getMutationExecutor(getRegularUser().getDomainUsername());
-      failAttachmentCreate(erinMutationExecutor, failedAttachmentInput);
+    final Boolean attachmentDisabled = (Boolean) attachmentSettings.get("featureDisabled");
+
+    if (!attachmentDisabled) {
```
Comment (gjvoosten) [inline src/test/java/mil/dds/anet/test/resources/AttachmentResourceTest.java:40]: Why is this now a conditional check? We know attachments are enabled in the dictionary used for testing, so the checks should be unconditional.

File: client/src/pages/reports/Form.js [hunk 6]
```diff
@@ -928,21 +928,23 @@ const ReportForm = ({
...
-                      edit={edit}
-                      saveRelatedObject={saveCallback}
-                      getRelatedObject={() => getRelatedObject(values)}
-                    />
-                  }
-                  onHandleBlur={() => {
-                    setFieldTouched("uploadAttachments", true, false)
-                  }}
-                />
+                {Settings.fields.attachment.featureDisabled && (
```
Comment (gjvoosten) [inline client/src/pages/reports/Form.js:931]: Shouldn't this be `!Settings.fields.attachment.featureDisabled && (…` i.e. the inverse condition?

File: src/main/resources/anet-schema.yml [hunk 7]
```diff
@@ -524,8 +524,12 @@ properties:
       attachment:
         type: object
         additionalProperties: false
-        required: [disabled, shortLabel, shortName, description, classification]
+        required: [featureDisabled, disabled, shortLabel, shortName, description, classification]
         properties:
+          featureDisabled:
+            type: boolean
+            title: Whether attachments are enabled
+            description: Used for enable/disable attachment feature
```
Comment (gjvoosten) [inline src/main/resources/anet-schema.yml:532]: Small grammar sugestion: "Used to enable/disable attachments",

File: src/test/java/mil/dds/anet/test/resources/AttachmentResourceTest.java [hunk 8]
```diff
@@ -34,181 +33,193 @@ public class AttachmentResourceTest extends AbstractResourceTest {
...
-        AttachmentInput.builder().withFileName("testDeleteAttachment.jpg").withMimeType(mimeType)
-            .withDescription("a test attachment created by testDeleteAttachment")
-            .withClassification(Classification.UNDEFINED)
-            .withAttachmentRelatedObjects(Collections.singletonList(testAroInput)).build();
-    final String createdAttachmentUuid =
-        succeedAttachmentCreate(adminMutationExecutor, testAttachmentInput);
-    assertThat(createdAttachmentUuid).isNotNull();
+    final Boolean attachmentDisabled = (Boolean) attachmentSettings.get("featureDisabled");
+
+    if (!attachmentDisabled) {
```
Comment (gjvoosten) [inline src/test/java/mil/dds/anet/test/resources/AttachmentResourceTest.java:107]: Why is this now a conditional check?

File: src/test/java/mil/dds/anet/test/resources/AttachmentResourceTest.java [hunk 9]
```diff
@@ -34,181 +33,193 @@ public class AttachmentResourceTest extends AbstractResourceTest {
...
-
-    succeedAttachmentUpdate(adminMutationExecutor,
-        getInput(updatedClassificationAttachment, AttachmentInput.class));
-    final Attachment updatedFilenameAttachment =
-        updatedClassificationReport.getAttachments().get(0);
-    assertThat(updatedFilenameAttachment.getFileName())
-        .isEqualTo(updatedClassificationAttachment.getFileName());
+    final Boolean attachmentDisabled = (Boolean) attachmentSettings.get("featureDisabled");
+
+    if (!attachmentDisabled) {
```
Comment (gjvoosten) [inline src/test/java/mil/dds/anet/test/resources/AttachmentResourceTest.java:162]: Why is this now a conditional check?

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: client/src/pages/reports/Show.js (hunk 2 above)
Comment (mk-simsoft) [reply client/src/pages/reports/Show.js:676]: https://github.com/NCI-Agency/anet/tree/AB-894-permission-control-to-upload-attachments I did it, github status have error so its not display here

File: client/src/pages/reports/Form.js (hunk 6 above)
Comment (mk-simsoft) [reply client/src/pages/reports/Form.js:931]: https://github.com/NCI-Agency/anet/tree/AB-894-permission-control-to-upload-attachments I did it, github status have error so its not display here

**Review by gjvoosten — APPROVED** [review]

**Pull request merged**